### PR TITLE
DMSwarm Example with Interpolation and Projection

### PR DESCRIPTION
**Pull request by jeremylt**
This basic example will have 2 main features

- [x] `DMSwarm[Get,Restore]ReferenceCoordinates` prototype
- [x] Comparing PETSc and libCEED interp at swarm points
- [x] Use new `ElemRestrictionApplyAtPointsInElement` 
- [x] Projection from swarm to mesh
- [ ] New CI build for updated PETSc

### CODE REVIEW

**Comment by jeremylt** [comment]
This will probably take a little longer than anticipated - this is slow going.

I think the basic logic is now in here though. Its ugly code that needs to be tidied though.

File: examples/petsc/dmswarm.c
Comment (jeremylt) [inline examples/petsc/dmswarm.c:92]: Maybe this? I'm still deciding exactly what to put in this index set.
```suggestion
      point_cell_numbers[cell_points[p]] = cell - cell_start;
```

Related - we should decide what sort of restriction operation we need to create for this and how to update it. Of note for this restriction is the fact that a) we have different numbers of swarm points per cell 2) we need to update the mapping or create/destroy this restriction once per time step

File: examples/petsc/dmswarm.c
Comment (jeremylt) [inline examples/petsc/dmswarm.c:55]: Note - I put `dm_mesh` as an argument here because there is no `DMSwarmGetCellDM`. This is fine since this will be an internal only method, but I think we should *also* add `DMSwarmGetCellDM` since that will be useful for us.

File: examples/petsc/dmswarm.c
Comment (jeremylt) [inline examples/petsc/dmswarm.c:112]: It is my understanding that using `PETSC_COMM_SELF` here means this will be local to the process?

File: examples/petsc/dmswarm.c
Comment (jeremylt) [inline examples/petsc/dmswarm.c:55]: ToDo - add PETSc style documentation header

File: examples/petsc/dmswarm.c
Comment (jeremylt) [inline examples/petsc/dmswarm.c:162]: These functions aren't linking correctly. Also, I don't really want to have to include the BP data just for the purpose of including it here. `CreateBasisFromPlex` should be updated to accept a `NULL` BP data and use standard quadrature in that case.

File: examples/petsc/dmswarm.c
Comment (jeremylt) [inline examples/petsc/dmswarm.c:318]: We do this internally in the CPU backends a lot. It might be nice to have a utility to make a single element evec or a 8 element batch evec?

File: examples/petsc/dmswarm.c
Comment (jeremylt) [inline examples/petsc/dmswarm.c:350]: Need to think more here - I use this pattern since I'm trying to only use the data available from the call to `DMSwarmCreateReferenceCoordinates` (the new function we're creating in this MR)

File: examples/petsc/dmswarm.c
Comment (jeremylt) [inline examples/petsc/dmswarm.c:296]: Variable naming is a bit of a thick soup right now. I'm iterating as I go to balance clarity, accuracy, and brevity.

File: examples/petsc/dmswarm.c
Comment (jedbrown) [reply examples/petsc/dmswarm.c:92]: For the particle restriction, two options come to mind:

1. Sort the particles by cell and provide the offsets `CeedInt particle_offset[num_cells + 1]`. These are similar to row starts in a CSR matrix format. The particles for each cell are contiguous in memory.
2. Create an integer permutation array `particle_index[num_particles]` so that the particles themselves don't need to be sorted. This requires less up-front work, but then a less regular memory access pattern.

I think either is fine, though I don't know if `DMSwarm` supports that intrusive sort (which has to also reorder every field). The second option is more flexible, especially if different operators need to work over different parts of the domain (so that the particles they need to work cannot be contiguous for each operator). That makes me lean toward 2.

File: examples/petsc/dmswarm.c
Comment (zatkins-dev) [inline examples/petsc/dmswarm.c:150]: Small note, I think this should be:
```suggestion
    PetscCall(PetscFECreateLagrange(comm, dim, dim, PETSC_FALSE, geometry_order, q_order, &fe_coord));
```

**Comment by zatkins-dev** [comment]
I think I found a bug in `CeedBasisApplyAtPoints` while trying to get this to run: #1287

**Comment by zatkins-dev** [comment]
I think I found a bug in `CeedBasisApplyAtPoints` while trying to get this to run: #1287

**Comment by jeremylt** [comment]
Example links and runs! I need to clean up the code, but it seems that the initial usage of DMSwarm is working as intended.

File: examples/petsc/dmswarm.c
Comment (zatkins-dev) [inline examples/petsc/dmswarm.c:219]: We should use `DMSwarmPICField_coor` instead of a string constant here.

File: examples/petsc/dmswarm.c
Comment (zatkins-dev) [inline examples/petsc/dmswarm.c:278]: Use `DMSwarmPICField_coor`

File: examples/petsc/dmswarm.c
Comment (zatkins-dev) [inline examples/petsc/dmswarm.c:325]: Use `DMSwarmPICField_coor`

File: examples/petsc/dmswarm.c
Comment (zatkins-dev) [inline examples/petsc/dmswarm.c]:
```suggestion
    PetscCall(DMSwarmGetField(dm_swarm, DMSwarmPICField_coor, NULL, NULL, (void **)&point_coords));
```

File: examples/petsc/dmswarm.c
Comment (zatkins-dev) [inline examples/petsc/dmswarm.c]:
```suggestion
    PetscCall(DMSwarmRestoreField(dm_swarm, DMSwarmPICField_coor, NULL, NULL, (void **)&coord_points));
```

File: examples/petsc/dmswarm.c
Comment (zatkins-dev) [inline examples/petsc/dmswarm.c]:
```suggestion
    PetscCall(DMSwarmRestoreField(dm_swarm, DMSwarmPICField_coor, NULL, NULL, (void **)&point_coords));
```

File: examples/petsc/dmswarm.c
Comment (zatkins-dev) [inline examples/petsc/dmswarm.c:81]:
```suggestion
  PetscCall(DMSwarmGetField(dm_swarm, DMSwarmPICField_coor, NULL, NULL, (void **)&coords_points_true));
```

File: examples/petsc/dmswarm.c
Comment (zatkins-dev) [inline examples/petsc/dmswarm.c:69]:
```suggestion
    PetscCall(DMSwarmCreateLocalVectorFromField(dm_swarm, DMSwarmPICField_coor, &true_coords));
    PetscCall(VecDuplicate(true_coords, ref_coords));
    PetscCall(DMSwarmDestroyLocalVectorFromField(dm_swarm, DMSwarmPICField_coor, &true_coords));
```

File: examples/petsc/dmswarm.c
Comment (zatkins-dev) [inline examples/petsc/dmswarm.c:108]:
```suggestion
  PetscCall(DMSwarmRestoreField(dm_swarm, DMSwarmPICField_coor, NULL, NULL, (void **)&coords_points_ref));
```

File: examples/petsc/dmswarm.c
Comment (zatkins-dev) [inline examples/petsc/dmswarm.c]:
```suggestion
    PetscCall(DMSwarmGetField(dm_swarm, DMSwarmPICField_coor, NULL, NULL, (void **)&coord_points));
```

**Review by zatkins-dev — CHANGES_REQUESTED** [review]
We should use the built-in constants for built-in fields

File: examples/petsc/dmswarm.c
Comment (zatkins-dev) [reply examples/petsc/dmswarm.c:56]: Isn't this `DMSwarmGetCellDM`? https://gitlab.com/petsc/petsc/-/blob/zach/bug-dm-swarm-tensor-permutation/src/dm/impls/swarm/swarm.c#L872

Comment (jeremylt) [reply examples/petsc/dmswarm.c:56]: dunno how I missed that

**Review by zatkins-dev — COMMENTED** [review]
empty comment

File: examples/petsc/dmswarm.c
Comment (zatkins-dev) [inline examples/petsc/dmswarm.c:260]: The sizes of the coordinate and local vectors aren't related like this in general, especially for mesh order > 1

File: examples/petsc/dmswarm.c
Comment (zatkins-dev) [inline examples/petsc/dmswarm.c:299]: This function seems to blow up (`coords_points_ref[0]` and `coords_points_ref[1]` in particular)

File: examples/petsc/dmswarm.c
Comment (jeremylt) [reply examples/petsc/dmswarm.c:260]: oh, we should project the coordinate space to match the solution space, for the purposes of this example

File: examples/petsc/dmswarm.c
Comment (zatkins-dev) [inline examples/petsc/dmswarm.c:303]: we aren't using the outputs of this call, that seems fishy

File: examples/petsc/dmswarm.c
Comment (jeremylt) [reply examples/petsc/dmswarm.c:260]: i.e. remove `geometry_order` and use `solution_order` in its place (or whatever I called those variables)

File: examples/petsc/dmswarm.c
Comment (jeremylt) [reply examples/petsc/dmswarm.c:303]: That's a placeholder for the grad checking code

File: examples/petsc/dmswarm.c
Comment (zatkins-dev) [reply examples/petsc/dmswarm.c:299]: replacing with this fixes the blow up
```suggestion
     PetscCall(DMPlexComputeCellGeometryFEM(dm_mesh, cell, NULL, v, J, invJ, &detJ));
      for (PetscInt p = 0; p < num_points_in_cell; ++p) {
        CoordinatesRealToRef(dim, dim, v0ref, v, invJ, &coord_points_true[p * dim], &coords_points_ref[p * dim]);
      }
```

File: examples/petsc/dmswarm.c
Comment (zatkins-dev) [inline examples/petsc/dmswarm.c:252]: Our coordinate ordering is wrong here, it should be:
```suggestion
      for (PetscInt d = 0; d < dim; d++) x[d] = x_array[i * dim + d];
```

File: examples/petsc/dmswarm.c
Comment (jeremylt) [inline examples/petsc/dmswarm.c:325]: Stray ++i instead of i++

File: examples/petsc/dmswarm.c
Comment (jeremylt) [reply examples/petsc/dmswarm.c]: As we dig more into this - I think what libCEED really needs is all of the indices of the points in the first local cell, followed by all the indices of the points in the second local cell, etc
something like
4
6
84
102
-1
23
35
76
-1
-1
46
78
...

where -1 indicates a break between cells

right now we have a nested loop with num_elem * num_local_points iterations all over the place, but this would bring us down to num_elem * ~num_points_per_elem, which is much, much fewer iterations

Comment (jeremylt) [reply examples/petsc/dmswarm.c]: resolved per conversation in Zulip - see changes in latest commit

**Comment by jeremylt** [comment]
Ready for review - merge blocked by https://gitlab.com/petsc/petsc/-/merge_requests/6843

**Comment by zatkins-dev** [comment]
> Ready for review - merge blocked by https://gitlab.com/petsc/petsc/-/merge_requests/6843

No longer blocked!

File: examples/petsc/dmswarm.c
Comment (zatkins-dev) [inline examples/petsc/dmswarm.c:272]:
```suggestion
. dm_swarm  - the `DMSwarm`

  Output Parameter:
+ is_points    - The IS object for indexing into points per cell
- X_points_ref - Vec holding the cell reference coordinates for local DMSwarm points
```
Minor: fix PETSc docstring parameter prefixes

**Comment by jeremylt** [comment]
Good point - we just need to update the PETSc builds on Noether. I got the Ratel ones, but I don't think I "own" the folders for the libCEED CI builds.

Really... we should get around to building containers, at least for our CPU builds. Not sure how to best also do containers for GPU builds (though I'd love for that to be an option for Ratel too)

File: examples/petsc/dmswarm.c
Comment (jeremylt) [reply examples/petsc/dmswarm.c:272]: I meant to ask about that - what's the magic decoder ring for what the . vs + vs - mean?

File: examples/petsc/dmswarm.c
Comment (zatkins-dev) [inline examples/petsc/dmswarm.c:270]:
```suggestion
  Input Parameter:
. dm_swarm  - the `DMSwarm`

  Output Parameters:
```
also plural/singular

File: examples/petsc/dmswarm.c
Comment (jedbrown) [reply examples/petsc/dmswarm.c]: In PETSc? That's a sowing idiosyncrasy.
https://wgropp.cs.illinois.edu/projects/software/sowing/doctext/node12.htm#Node12

**Comment by zatkins-dev** [comment]
Are we planning on implementing the other direction, i.e. points to cells, in this example?

**Comment by jeremylt** [comment]
Eventually. I think we can incrementally add functionality as our libCEED infrastructure matures this month.

**Comment by jeremylt** [comment]
I plan to update this with #1333 

If we add this to the v0.12 release, we can use it to help paint the picture that we are moving to MPM. Then we can refine this example when we add the CeedOperator with *AtPoints interface (hopefully by the end of the month, or at least a prototype version)

File: examples/petsc/dmswarm.c
Comment (zatkins-dev) [inline examples/petsc/dmswarm.c:21]: Can we add either a second test or just an example invocation for the lumped mass case? And maybe label this one as KSP?

File: examples/petsc/dmswarm.c
Comment (zatkins-dev) [inline examples/petsc/dmswarm.c:119]: The `-points` option (and its default) are somewhat unclear. Can we make this a `-points_per_cell` option (or otherwise support that)?

File: examples/petsc/dmswarm.c
Comment (zatkins-dev) [inline examples/petsc/dmswarm.c:120]:
```suggestion
  PetscCall(PetscOptionsScalar("-tolerance", "Tolerance for swarm point values and projection relative L2 error", NULL, tolerance, &tolerance, NULL));
```
Should make it clear that we are using this differently for point values and projection error

File: examples/petsc/dmswarm.c
Comment (zatkins-dev) [inline examples/petsc/dmswarm.c:96]: In an ideal world this should be an enum, but I don't think it really matter for this example.

File: examples/petsc/dmswarm.c
Comment (zatkins-dev) [inline examples/petsc/dmswarm.c:46]: I think we should revisit this struct when we integrate with Ratel -- parts of this should be stored on the whole mesh, but other parts are field-specific. Maybe something like this for handling multiple fields? Not sure how we will want to minimize redundant work computing particles in cells. 
```c
struct DMSwarmCeedFieldContext_ {
  CeedElemRestriction restriction_u_mesh, restriction_u_points;
  CeedBasis           basis_u;
  const char          fieldname[MAX_NAME_LENGTH];
}
```

**Review by zatkins-dev — APPROVED** [review]
Mostly small comments and thoughts about moving forward. I would be fine with merging in its current state.

File: examples/petsc/dmswarm.c
Comment (jeremylt) [reply examples/petsc/dmswarm.c:21]: Added in the lastest - good idea

File: examples/petsc/dmswarm.c
Comment (jeremylt) [reply examples/petsc/dmswarm.c:48]: Definitely - I think we'll ultimately store this into an operator field so we don't have to create a special struct

File: examples/petsc/dmswarm.c
Comment (jeremylt) [reply examples/petsc/dmswarm.c:98]: Yeah. My main hangup is that I couldn't figure out what to call the other version of the points distribution? (Though we don't really use it much currently it is nice having the option to have a different number of points in each cell to see what that does to accuracy)

File: examples/petsc/dmswarm.c
Comment (jeremylt) [reply examples/petsc/dmswarm.c]: Added a `-points_per_cell` option

File: examples/petsc/dmswarm.c
Comment (zatkins-dev) [reply examples/petsc/dmswarm.c:99]: That's fair, maybe like "path" or "semicircle"?

**Comment by zatkins-dev** [comment]
Will we want to update the `junit.py` to ensure this is only run with CPU backends?

**Comment by jeremylt** [comment]
`ApplyBlock` should give a "Not supported" message, so it should error out ok

File: examples/petsc/dmswarm.c
Comment (zatkins-dev) [inline examples/petsc/dmswarm.c:528]:
```suggestion
    PetscBool user_set_points_per_cell = PETSC_FALSE;
```

File: examples/petsc/dmswarm.c
Comment (zatkins-dev) [inline examples/petsc/dmswarm.c:531]:
```suggestion
    PetscBool user_set_points_per_cell = PETSC_FALSE;
    PetscInt  dim = 3, points_per_cell = num_points;
    PetscInt  num_cells[] = {1, 1, 1};
    DM        dm_mesh;
```

File: examples/petsc/dmswarm.c
Comment (zatkins-dev) [inline examples/petsc/dmswarm.c:569]: `points_per_cell_1d` is scoped wrong here

**Review by jedbrown — APPROVED** [review]
LGTM. We've run and visualized results here, and it seems correct.

**Comment by jeremylt** [comment]
@jedbrown can you rebuild PETSc for CI? I can't
```
git@gitlab.com: Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

**Comment by jedbrown** [comment]
Done.

**Comment by jeremylt** [comment]
CI is running. I'll squash+merge when it finishes.

**Pull request merged**
